pkg/controller/daemon/update.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package daemon
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	intstrutil "k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/kubernetes/pkg/api/v1"
+	extensions "k8s.io/kubernetes/pkg/apis/extensions/v1beta1"
+	"k8s.io/kubernetes/pkg/controller/daemon/util"
+	podutil "k8s.io/kubernetes/pkg/controller/deployment/util"
+)
+
+func (dsc *DaemonSetsController) rollingUpdate(ds *extensions.DaemonSet) error {
+	newPods, oldPods, err := dsc.getAllDaemonSetPods(ds)
+	allPods := append(oldPods, newPods...)
+
+	maxUnavailable, numUnavailable, err := dsc.GetUnavailableNumbers(ds, allPods)
+	if err != nil {
+		glog.Errorf("Couldn't get unavailable numbers: %v", err)
+		return nil
+	}
+	oldAvailablePods, oldUnavailablePods := util.SplitByAvailablePods(ds, oldPods)
+
+	// for oldPods delete all not running pods
+	var podsToDelete []string
+	glog.V(4).Infof("Marking all unavailable old pods for deletion")
+	for _, pod := range oldUnavailablePods {
```

Comment:
Yes, only _old_ unavailable pods are deleted here. @kargakis are you suggesting it's safer to just kill old pods that are not running (i.e. don't kill the running but unavailable old pods)? Or we don't care since it's old?